content/reference/v1beta1/repeatable.md

```diff
@@ -0,0 +1,284 @@
+
+## Templating
+
+The repeat items are called with the delimeters `repl[[ .itemName ]]` or `[[repl .itemName ]]`.  These delimiters can be placed anywhere inside of the `yamlPath` target node.
+```yaml
+    - port: '{{repl ConfigOption "[[repl .service_ports ]]" | ParseInt }}'
+```
+This repeatable templating is not compatible with sprig templating functions.  It is designed for inserting repeatable `keys` into the manifest. Repeatable templating **can** be placed inside of Replicated config templating.
+
+## Ordering
+
+Repeatable templates are processed before config template rendering.
+
+Repeatable items are processed in order of the template targets in the Config Spec file.  Effectively, this ordering is from the top of the Config Spec, by Config Group, by Config Item, and then by template target.
+
+```yaml
+    - name: ports
+      items:
+      - name: service_ports
+        title: Service Ports
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: v1 #processed first
+          kind: Service
+          name: my-service
+          namespace: my-app
+          yamlPath: 'spec.ports[0]'
+        - apiVersion: v1 #processed second
+          kind: Service
+          name: my-service
+          namespace: my-app
+          yamlPath:
+    {... next item}
+      - name: other_ports
+        title: Other Service Ports
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: v1 #processed third
+          kind: Service
+          name: my-other-service
+          namespace: my-app
+    {... next group}
+    - name: deployments
+      items:
+      - name: deployment-name
+        title: Deployment Names
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: apps/v1 #processed fourth
+          kind: Deployment
+          name: my-deployment
+          namespace: my-app
+          yamlPath:
+    {... remainder of the Config Spec}
+```
+
+## Repeatable Item Example for a YamlPath
+**Config spec**
+```yaml
+    - name: ports
+      items:
+      - name: service_ports
+        title: Service Ports
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: v1
+          kind: Service
+          name: my-service
+          namespace: my-app
+          yamlPath: 'spec.ports[0]'
+        valuesByGroup:
+          ports:
+            port-default-1: "80"
+            port-default-2: "443"
+```
+**Template manifest**
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+  namespace: my-app
+spec:
+  type: NodePort
+  ports:
+  - port: '{{repl ConfigOption "[[repl .service_ports ]]" | ParseInt }}'
+    name: '{{repl ConfigOptionName "[[repl .service_ports ]]" }}'
+  selector:
+    app: repeat_example
+    component: my-deployment
+```
+
+**After repeatable config processing**
```

Comment:
After our comments in the other review, I realized the generated names don't make an appearance here; everything comes from the config spec. It would be helpful to flesh out the example more and explain that the user dynamically adds another value, a name is generated, and that gets plugged into the template manifest. Otherwise, it feels like this could all be done without repeatable even being necessary (the template provided all the information, the user didn't change anything).

What if we removed port 443 from the config and just had one default, port 80. Then we describe that the user added a second port, 443. Then we have a generated name (whatever that would be, nginx-xxxx from your demo the other day) that gets plugged in for the resulting manifest. That would help me see the dynamic nature here and better understand why this is necessary and how a user's input is reflected in the result. And then do something similar for the YAML document example too.

What do you think?

Reply:
Solid idea.  I've updated the full example at the bottom to say port 80 is the shipped default, and port 443 has been added by the end-user.  Let me know your thoughts on the new example.

Reply:
This can resolve, though I made a couple comments in a different review on this topic.